File: bindings/java/src/Transaction.c
// Copyright (c) 2025-present The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "Transaction.h"

#include <btck/btck.h>

#include <jni.h>
#include <stdint.h>
#include <stdlib.h>

static jfieldID get_impl_field(JNIEnv* env, jobject obj)
{
  jclass cls = (*env)->GetObjectClass(env, obj);
  return (*env)->GetFieldID(env, cls, "impl", "J");
}

jlong Java_Transaction_construct(JNIEnv* env, jclass clazz, jobject raw)
{
  (void)clazz;
  if (raw == NULL) {
    return 0;
  }

  void* data = (*env)->GetDirectBufferAddress(env, raw);
  jlong len = (*env)->GetDirectBufferCapacity(env, raw);
  if (data == NULL || len <= 0) {
    return 0;
  }

  return (jlong)(uintptr_t)BtcK_Transaction_New(data, (size_t)len, NULL);
}

void Java_Transaction_close(JNIEnv* env, jobject self)
{
  jfieldID fid = get_impl_field(env, self);
  jlong handle = (*env)->GetLongField(env, self, fid);
  struct BtcK_Transaction* ptr = (struct BtcK_Transaction*)(uintptr_t)handle;
  if (ptr != NULL) {
    BtcK_Transaction_Free(ptr);
    (*env)->SetLongField(env, self, fid, (jlong)0);
  }
}

jint Java_Transaction_size(JNIEnv* env, jobject self)
{
  jfieldID fid = get_impl_field(env, self);
  jlong handle = (*env)->GetLongField(env, self, fid);
  struct BtcK_Transaction* ptr = (struct BtcK_Transaction*)(uintptr_t)handle;
  if (ptr == NULL) {
    return 0;
  }

  return (jint)BtcK_Transaction_CountOutputs(ptr);
}

jobject Java_Transaction_get(JNIEnv* env, jobject self, jint idx)
{
  jfieldID fid = get_impl_field(env, self);
  jlong handle = (*env)->GetLongField(env, self, fid);
  struct BtcK_Transaction* ptr = (struct BtcK_Transaction*)(uintptr_t)handle;
  if (ptr == NULL) {
    return NULL;
  }

  struct BtcK_TransactionOutput* out =
    BtcK_Transaction_GetOutput(ptr, (size_t)idx, NULL);
  if (out == NULL) {
    return NULL;
  }

  jclass outClass = (*env)->FindClass(env, "TransactionOutput");
  if (outClass == NULL) {
    return NULL;
  }

  jmethodID ctor = (*env)->GetMethodID(env, outClass, "<init>", "(J)V");
  if (ctor == NULL) {
    return NULL;
  }

  jobject outObj =
    (*env)->NewObject(env, outClass, ctor, (jlong)(uintptr_t)out);
  return outObj;
}

jobject Java_Transaction_asByteBuffer(JNIEnv* env, jobject self)
{
  jfieldID fid = get_impl_field(env, self);
  jlong handle = (*env)->GetLongField(env, self, fid);
  struct BtcK_Transaction* ptr = (struct BtcK_Transaction*)(uintptr_t)handle;
  if (ptr == NULL) {
    return NULL;
  }

  size_t len = 0;
  void const* data = BtcK_Transaction_AsBytes(ptr, &len);
  if (data == NULL || len == 0) {
    return NULL;
  }

  return (*env)->NewDirectByteBuffer(env, (void*)data, (jlong)len);
}

File: bindings/swift/transaction.swift
// Copyright (c) 2025-present The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

import Foundation
@_implementationOnly import btck_c

public protocol TransactionProtocol {
    var ptr: OpaquePointer { get }
}

public extension TransactionProtocol { // RandomAccessCollection {
    typealias Index = Int
    typealias Element = TransactionOutputRef

    var startIndex: Int { 0 }
    var endIndex: Int { BtcK_Transaction_CountOutputs(ptr) }

    func index(after i: Int) -> Int { i + 1 }
    func index(before i: Int) -> Int { i - 1 }

    subscript(position: Int) -> TransactionOutputRef {
        TransactionOutputRef(ptr: BtcK_Transaction_GetOutput(ptr, position))
    }
}

public extension TransactionProtocol { // CustomStringConvertible {
    var description: String {
        let len = BtcK_Transaction_ToString(ptr, nil, 0)
        guard len > 0 else { return "" }
        let buf = UnsafeMutablePointer<CChar>.allocate(capacity: Int(len) + 1)
        defer { buf.deallocate() }
        let written = BtcK_Transaction_ToString(ptr, buf, Int(len) + 1)
        guard written >= 0 else { return "" }
        return String(cString: buf)
    }
}

public class Transaction: TransactionProtocol, RandomAccessCollection, CustomStringConvertible {
    public internal(set) var ptr: OpaquePointer

    public init(raw: Data) throws {
        var err: OpaquePointer? = nil
        let ptr = raw.withUnsafeBytes({ buf in
            BtcK_Transaction_New(buf.baseAddress, raw.count, &err)
        })
        if let err {
            throw BtcKError(err)
        }
        self.ptr = ptr!
    }

    internal init(ptr: OpaquePointer) {
        self.ptr = ptr
    }

    deinit {
        BtcK_Transaction_Free(ptr)
    }
}

public class TransactionRef: TransactionProtocol, RandomAccessCollection, CustomStringConvertible {
    public let ptr: OpaquePointer

    internal init(ptr: OpaquePointer) {
        self.ptr = ptr
    }
}

public extension TransactionProtocol {
    var data: Data {
        var buffer = Data()
        let callback: @convention(c) (UnsafeRawPointer?, size_t, UnsafeMutableRawPointer?) -> Int32 = { bytes, sz, userdata in
            guard let bytes, let userdata else { return -1 }
            let dataPtr = userdata.assumingMemoryBound(to: Data.self)
            dataPtr.pointee.append(bytes.assumingMemoryBound(to: UInt8.self), count: sz)
            return 0
        }
        let result = BtcK_Transaction_ToBytes(ptr, callback, &buffer)
        guard result == 0 else { return Data() }
        return buffer
    }
}
